# <u>**ACADEMY**</u> : Educational Systems Data Analysis

<details open>
<summary><u style="font-size: 2rem">Summary: </u></summary>

* <a href="#author" data-toc-modified-id="author" >Author</a>
* <a href="#introduction" data-toc-modified-id="introduction" >Introduction</a>
* <a href="#functions" data-toc-modified-id="functions" >Functions</a>
* <a href="#import-necessary-libraries" data-toc-modified-id="import-necessary-libraries" >Import Necessary Libraries</a>
* <a href="#load-the-data" data-toc-modified-id="load-the-data" >Load the Data</a>
* <a href="#initial-data-exploration" data-toc-modified-id="initial-data-exploration">Initial Data Exploration</a>
    * <a href="#stats-data" data-toc-modified-id="stats-data">1. EdStatsData</a>
* <a href="#data-cleaning" data-toc-modified-id="data-cleaning">Data Cleaning</a>



</details>

<a id='author'></a>

## Author
> Mohamed Ali EL HAMECH 

- @MasterCodeDevelop (Github)
- E-mail : master.code.develop@gmail.com
- Affiliation : ACADEMY #OpenClassRooms
<p><img align="left" src="https://github-readme-stats.vercel.app/api/top-langs?username=mastercodedevelop&show_icons=true&locale=en&layout=compact" alt="mastercodedevelop" />

<a id='introduction'></a>

## Introduction:
>In the rapidly evolving landscape of online education, ACADEMY, a burgeoning EdTech start-up, has been making significant strides by offering high-quality online training content tailored for high school and university students. As part of its strategic vision, ACADEMY is keenly exploring opportunities for international expansion, aiming to tap into markets with a robust educational framework and a potential clientele for its services.

_The objective of this analysis is twofold. Firstly, it seeks to delve into global educational data, sourced from the World Bank, to ascertain the viability and potential of various countries as prospective markets for ACADEMY. This dataset, curated by the "EdStats All Indicator Query" of the World Bank, boasts a comprehensive collection of over 4,000 international indicators. These indicators span a range of metrics, from access to education and graduation rates to insights about educators and educational expenditures._

_Secondly, this analysis aims to provide a clear, data-driven narrative that would aid ACADEMY's decision-makers in charting the company's international trajectory. By evaluating the quality of the dataset, understanding its breadth and depth, and extracting relevant insights, we hope to offer a roadmap that aligns with ACADEMY's mission and vision._

<a id='import-necessary-libraries'></a>

## Import Necessary Libraries
> Before you can work with the data, you need to import the necessary libraries.

In [352]:
import pandas as pd
import numpy as np

<a id='functions'></a>

## Functions

In [353]:
def missing_values_table(df):
    """
    Return a DataFrame displaying the count and percentage of missing values for each column.
    
    Parameters:
    - df (DataFrame): The input DataFrame to check for missing values.
    
    Returns:
    - DataFrame: A DataFrame with columns 'Missing Values' and 'Missing Percentages', sorted in descending order of missing percentages.
    """
    
    # Calculate missing values and their percentages
    missing_values = df.isnull().sum()
    missing_percentages = (df.isnull().mean() * 100).round(2)  # rounding to 2 decimal places for clarity
    
    # Create a DataFrame to display missing data info
    missing_data = pd.DataFrame({
        'Missing Values': missing_values, 
        'Missing Percentages (%)': missing_percentages
    })
    
    # Filter out columns with no missing values and sort by percentage
    missing_data = missing_data[missing_data['Missing Values'] > 0].sort_values(by='Missing Percentages (%)', ascending=False)
    
    return missing_data


<a id='load-the-data'></a>

## Load the Data
> Importing the dataset into the environment for analysis and exploration.

In [354]:
# These are the basic data variables we will be working with
COUNTRY_SERIES = pd.read_csv('./data/EdStatsCountry-Series.csv')
COUNTRY = pd.read_csv('./data/EdStatsCountry.csv')
DATA = pd.read_csv('./data/EdStatsData.csv')
FOOT_NOTE = pd.read_csv('./data/EdStatsFootNote.csv')
EdStatsSeries = pd.read_csv('./data/EdStatsSeries.csv')

# These are the data variables we will be working with and can be modified
country_series = COUNTRY_SERIES.copy()
country = COUNTRY.copy()
data = DATA.copy()
foot_note = FOOT_NOTE.copy()


<a id='initial-data-exploration'></a>

## Initial Data Exploration
> Diving into the dataset to understand its structure, content, and characteristics.

<a id='stats-data'></a>

### 1 - <u>EdStatsData :</u>

#### 1.1 - Preview of the First and Last Rows of the Dataset
> View of the dataset's start & end, providing a quick data structure glimpse.

In [355]:
print("First 5 rows of the dataset:")
DATA.head()

First 5 rows of the dataset:


Country Name Country Code  \
0   Arab World          ARB   
1   Arab World          ARB   
2   Arab World          ARB   
3   Arab World          ARB   
4   Arab World          ARB   

                                      Indicator Name  Indicator Code  \
0  Adjusted net enrolment rate, lower secondary, ...      UIS.NERA.2   
1  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.F   
2  Adjusted net enrolment rate, lower secondary, ...  UIS.NERA.2.GPI   
3  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.M   
4  Adjusted net enrolment rate, primary, both sex...     SE.PRM.TENR   

        1970       1971       1972       1973       1974      1975  ...  2060  \
0        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
1        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
2        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
3        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
4  54.822121  54.894138  56.209438  57.267109  57.991138  59.36554  ...   NaN   

   2065  2070  2075  2080  2085  2090  2095  2100  Unnamed: 69  
0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
1   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
2   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
4   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  

[5 rows x 70 columns]

In [356]:
print("Last 5 rows of the dataset:")
print(DATA.tail())

Last 5 rows of the dataset:
       Country Name Country Code  \
886925     Zimbabwe          ZWE   
886926     Zimbabwe          ZWE   
886927     Zimbabwe          ZWE   
886928     Zimbabwe          ZWE   
886929     Zimbabwe          ZWE   

                                           Indicator Name  \
886925  Youth illiterate population, 15-24 years, male...   
886926  Youth literacy rate, population 15-24 years, b...   
886927  Youth literacy rate, population 15-24 years, f...   
886928  Youth literacy rate, population 15-24 years, g...   
886929  Youth literacy rate, population 15-24 years, m...   

              Indicator Code  1970  1971  1972  1973  1974  1975  ...  2060  \
886925      UIS.LP.AG15T24.M   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
886926     SE.ADT.1524.LT.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
886927  SE.ADT.1524.LT.FE.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
886928  SE.ADT.1524.LT.FM.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...   

> <u>Note:</u> By examining the first and last 5 lines, we notice a significant number of missing data, marked as 'NaN'.

#### 1.2 - General information about the DataFrame

In [357]:
# This includes the data type, number of non-zero values, etc.
print(DATA.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 886930 entries, 0 to 886929
Data columns (total 70 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   Country Name    886930 non-null  object 
 1   Country Code    886930 non-null  object 
 2   Indicator Name  886930 non-null  object 
 3   Indicator Code  886930 non-null  object 
 4   1970            72288 non-null   float64
 5   1971            35537 non-null   float64
 6   1972            35619 non-null   float64
 7   1973            35545 non-null   float64
 8   1974            35730 non-null   float64
 9   1975            87306 non-null   float64
 10  1976            37483 non-null   float64
 11  1977            37574 non-null   float64
 12  1978            37576 non-null   float64
 13  1979            36809 non-null   float64
 14  1980            89122 non-null   float64
 15  1981            38777 non-null   float64
 16  1982            37511 non-null   float64
 17  1983      

In [358]:
# Display data types of each column
print(DATA.dtypes)

# Count columns by their data type
dtype_counts = DATA.dtypes.value_counts()
print("\nNumber of columns by data type:")
print(dtype_counts)


Country Name       object
Country Code       object
Indicator Name     object
Indicator Code     object
1970              float64
                   ...   
2085              float64
2090              float64
2095              float64
2100              float64
Unnamed: 69       float64
Length: 70, dtype: object

Number of columns by data type:
float64    66
object      4
dtype: int64


> <u>Note :</u>  While examining the data types of the columns, we observe that the first 4 columns are of type object, indicating they contain strings. Whereas columns with years as titles are of type float64, signifying they hold numerical data.

#### 1.3 - Columns

In [359]:
# Display columns names
print(DATA.columns)

Index(['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code',
       '1970', '1971', '1972', '1973', '1974', '1975', '1976', '1977', '1978',
       '1979', '1980', '1981', '1982', '1983', '1984', '1985', '1986', '1987',
       '1988', '1989', '1990', '1991', '1992', '1993', '1994', '1995', '1996',
       '1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004', '2005',
       '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013', '2014',
       '2015', '2016', '2017', '2020', '2025', '2030', '2035', '2040', '2045',
       '2050', '2055', '2060', '2065', '2070', '2075', '2080', '2085', '2090',
       '2095', '2100', 'Unnamed: 69'],
      dtype='object')


> <u>Note :</u> When observing the columns, we first notice that the titles of the first 4 columns are words. Then, starting from the fifth column, the titles represent years, ranging from 1970 to 2100 in ascending order. From 1970 to 2017, the years increase by 1. There are no columns for the years between 2017 and 2020, while from 2020 to 2100, the years increase in intervals of 5. Lastly, there's a column named 'Unnamed: 69' which appears to be there by mistake.

#### 1.4 - Missing Values & Percentages
> Overview of missing data in columns, presented as counts & percentages

In [360]:
print("\nNumber of missing values for each column:")
DATA.isnull()


Number of missing values for each column:


Country Name  Country Code  Indicator Name  Indicator Code   1970  \
0              False         False           False           False   True   
1              False         False           False           False   True   
2              False         False           False           False   True   
3              False         False           False           False   True   
4              False         False           False           False  False   
...              ...           ...             ...             ...    ...   
886925         False         False           False           False   True   
886926         False         False           False           False   True   
886927         False         False           False           False   True   
886928         False         False           False           False   True   
886929         False         False           False           False   True   

         1971   1972   1973   1974   1975  ...  2060  2065  2070  2075  2080  \
0        True   True   True   True   True  ...  True  True  True  True  True   
1        True   True   True   True   True  ...  True  True  True  True  True   
2        True   True   True   True   True  ...  True  True  True  True  True   
3        True   True   True   True   True  ...  True  True  True  True  True   
4       False  False  False  False  False  ...  True  True  True  True  True   
...       ...    ...    ...    ...    ...  ...   ...   ...   ...   ...   ...   
886925   True   True   True   True   True  ...  True  True  True  True  True   
886926   True   True   True   True   True  ...  True  True  True  True  True   
886927   True   True   True   True   True  ...  True  True  True  True  True   
886928   True   True   True   True   True  ...  True  True  True  True  True   
886929   True   True   True   True   True  ...  True  True  True  True  True   

        2085  2090  2095  2100  Unnamed: 69  
0       True  True  True  True         True  
1       True  True  True  True         True  
2       True  True  True  True         True  
3       True  True  True  True         True  
4       True  True  True  True         True  
...      ...   ...   ...   ...          ...  
886925  True  True  True  True         True  
886926  True  True  True  True         True  
886927  True  True  True  True         True  
886928  True  True  True  True         True  
886929  True  True  True  True         True  

[886930 rows x 70 columns]

In [361]:
missing_values_table(DATA)

Missing Values  Missing Percentages (%)
Unnamed: 69          886930                   100.00
2017                 886787                    99.98
2016                 870470                    98.14
1971                 851393                    95.99
1973                 851385                    95.99
...                     ...                      ...
2011                 740918                    83.54
2012                 739666                    83.40
2000                 710254                    80.08
2005                 702822                    79.24
2010                 644488                    72.67

[66 rows x 2 columns]

In [362]:
# Display missing data > 70%
missing_data_70 = missing_values_table(DATA)[missing_values_table(DATA)['Missing Percentages (%)'] > 70]
print("Display missing data > 70% :")
print(missing_data_70)
print( f"total columns that have missing data > 70% :", missing_data_70.index.size)

Display missing data > 70% :
             Missing Values  Missing Percentages (%)
Unnamed: 69          886930                   100.00
2017                 886787                    99.98
2016                 870470                    98.14
1971                 851393                    95.99
1973                 851385                    95.99
...                     ...                      ...
2011                 740918                    83.54
2012                 739666                    83.40
2000                 710254                    80.08
2005                 702822                    79.24
2010                 644488                    72.67

[66 rows x 2 columns]
total columns that have missing data > 70% : 66


In [363]:
# Display missing data > 90%
missing_data_90 = missing_values_table(DATA)[missing_values_table(DATA)['Missing Percentages (%)'] > 90]
print("Display missing data > 90% :")
print(missing_data_90)
print( f"total columns that have missing data > 90% :", missing_data_90.index.size)

Display missing data > 90% :
             Missing Values  Missing Percentages (%)
Unnamed: 69          886930                   100.00
2017                 886787                    99.98
2016                 870470                    98.14
1971                 851393                    95.99
1973                 851385                    95.99
1972                 851311                    95.98
1974                 851200                    95.97
1979                 850121                    95.85
1982                 849419                    95.77
1989                 849390                    95.77
1976                 849447                    95.77
1977                 849356                    95.76
1978                 849354                    95.76
1983                 848470                    95.66
1988                 848378                    95.65
1984                 848324                    95.65
1987                 848289                    95.64
1981             

> <u>Note :</u> Out of the 70 columns, a significant number show missing data: 44 columns have over 90% missing data, and 65 have more than 70% missing. Notably, the column "Unnamed: 69" is entirely missing data, at a full 100%.

#### 1.5 - Descriptive Statistics
> Displaying descriptive statistics to understand central tendencies, dispersion, and shape of the dataset's distribution.

In [364]:
DATA.describe()

1970          1971          1972          1973          1974  \
count  7.228800e+04  3.553700e+04  3.561900e+04  3.554500e+04  3.573000e+04   
mean   1.974772e+09  4.253638e+09  4.592365e+09  5.105006e+09  5.401493e+09   
std    1.211687e+11  1.804814e+11  1.914083e+11  2.059170e+11  2.112150e+11   
min   -1.435564e+00 -1.594625e+00 -3.056522e+00 -4.032582e+00 -4.213563e+00   
25%    8.900000e-01  8.853210e+00  9.240920e+00  9.595200e+00  9.861595e+00   
50%    6.317724e+00  6.316240e+01  6.655139e+01  6.969595e+01  7.087760e+01   
75%    6.251250e+01  5.655200e+04  5.863650e+04  6.202900e+04  6.383675e+04   
max    1.903929e+13  1.986457e+13  2.100916e+13  2.238367e+13  2.282991e+13   

               1975          1976          1977          1978          1979  \
count  8.730600e+04  3.748300e+04  3.757400e+04  3.757600e+04  3.680900e+04   
mean   2.314288e+09  5.731808e+09  6.124437e+09  6.671489e+09  7.436724e+09   
std    1.375059e+11  2.215546e+11  2.325489e+11  2.473986e+11  2.660957e+11   
min   -3.658569e+00 -2.950945e+00 -3.174870e+00 -3.558749e+00 -2.973612e+00   
25%    1.400000e+00  9.312615e+00  9.519913e+00  1.000000e+01  1.000000e+01   
50%    9.677420e+00  7.101590e+01  7.133326e+01  7.290512e+01  7.510173e+01   
75%    7.854163e+01  5.682800e+04  5.739175e+04  5.940425e+04  6.411500e+04   
max    2.300634e+13  2.424128e+13  2.521383e+13  2.622101e+13  2.730873e+13   

       ...          2060          2065          2070          2075  \
count  ...  5.143600e+04  5.143600e+04  5.143600e+04  5.143600e+04   
mean   ...  7.224868e+02  7.271290e+02  7.283779e+02  7.266484e+02   
std    ...  2.215845e+04  2.287990e+04  2.352338e+04  2.408149e+04   
min    ... -1.630000e+00 -1.440000e+00 -1.260000e+00 -1.090000e+00   
25%    ...  3.000000e-02  3.000000e-02  2.000000e-02  2.000000e-02   
50%    ...  2.300000e-01  2.300000e-01  2.300000e-01  2.300000e-01   
75%    ...  7.505000e+00  7.500000e+00  7.300000e+00  7.100000e+00   
max    ...  2.951569e+06  3.070879e+06  3.169711e+06  3.246239e+06   

               2080          2085          2090          2095          2100  \
count  5.143600e+04  5.143600e+04  5.143600e+04  5.143600e+04  5.143600e+04   
mean   7.228327e+02  7.176899e+02  7.113072e+02  7.034274e+02  6.940296e+02   
std    2.455897e+04  2.496587e+04  2.530183e+04  2.556069e+04  2.574189e+04   
min   -9.200000e-01 -7.800000e-01 -6.500000e-01 -5.500000e-01 -4.500000e-01   
25%    1.000000e-02  1.000000e-02  1.000000e-02  1.000000e-02  1.000000e-02   
50%    2.300000e-01  2.300000e-01  2.300000e-01  2.300000e-01  2.200000e-01   
75%    6.722500e+00  6.080000e+00  5.462500e+00  4.680000e+00  4.032500e+00   
max    3.301586e+06  3.337871e+06  3.354746e+06  3.351887e+06  3.330484e+06   

       Unnamed: 69  
count          0.0  
mean           NaN  
std            NaN  
min            NaN  
25%            NaN  
50%            NaN  
75%            NaN  
max            NaN  

[8 rows x 66 columns]

> <u>Note :</u> By examining the standard deviation and quantiles for each column, it's evident that the average (mean), minimum (min), and maximum (max) values show significant variability across the different columns.

#### 1.6 - Display duplicate rows

In [365]:
duplicated_rows = DATA.duplicated().sum()
print(f"total duplicate rows:  {duplicated_rows}")

total duplicate rows:  0


> <u>Note :</u> We can observe that there are no duplicated rows.

#### 1.7 - Unique Values Check
> Checking unique values for categorical columns to understand the different categories present.

In [366]:
for column in DATA.select_dtypes(include=['object']).columns:
    print(f"Unique values for {column}: {DATA[column].nunique()}")
    print(DATA[column].unique())
    print("\n")

Unique values for Country Name: 242
['Arab World' 'East Asia & Pacific'
 'East Asia & Pacific (excluding high income)' 'Euro area'
 'Europe & Central Asia' 'Europe & Central Asia (excluding high income)'
 'European Union' 'Heavily indebted poor countries (HIPC)' 'High income'
 'Latin America & Caribbean'
 'Latin America & Caribbean (excluding high income)'
 'Least developed countries: UN classification' 'Low & middle income'
 'Low income' 'Lower middle income' 'Middle East & North Africa'
 'Middle East & North Africa (excluding high income)' 'Middle income'
 'North America' 'OECD members' 'South Asia' 'Sub-Saharan Africa'
 'Sub-Saharan Africa (excluding high income)' 'Upper middle income'
 'World' 'Afghanistan' 'Albania' 'Algeria' 'American Samoa' 'Andorra'
 'Angola' 'Antigua and Barbuda' 'Argentina' 'Armenia' 'Aruba' 'Australia'
 'Austria' 'Azerbaijan' 'Bahamas, The' 'Bahrain' 'Bangladesh' 'Barbados'
 'Belarus' 'Belgium' 'Belize' 'Benin' 'Bermuda' 'Bhutan' 'Bolivia'
 'Bosnia and Herze

> <u>Note: </u> We observe the presence of country names, regions, and indicator code

#### 1.8 - Dataset Dimensions & Unique Entries Overview
> - the number of rows 
> - the number of rows columns
> - unique number of countries 
> - unique number of indicators

In [367]:
print(f"Number of rows: {DATA.shape[0]}")
print(f"Number of columns {DATA.shape[1]}")
print(f"Nimber of unique countries: {DATA['Country Name'].nunique()}")
print(f"Number of unique indicators: {DATA['Indicator Name'].nunique()}")

Number of rows: 886930
Number of columns 70
Nimber of unique countries: 242
Number of unique indicators: 3665


> <u>Note :</u> The dataset, spanning 886 930 rows and 70 columns, encompasses 242 countries and 3,665 unique indicators. Its vast scope suggests in-depth information for each country and indicator, potentially over multiple years. The diversity of countries and indicators attests to a wealth of information suitable for varied analyses. This composition implies a global perspective with multiple data points per country.

#### 1.9 - <u>Notes</u> : 

The dataset, consisting of 886,930 rows and 70 columns, covers 242 countries with 3,665 unique indicators. The first four columns contain strings, while the subsequent ones, titled by years, hold numerical data. There's an annual progression from 1970 to 2017, a gap between 2017 and 2020, and then a five-year progression up to 2100. Many missing data points are observed, especially in the 'Unnamed: 69' column. Data variability is significant, and no rows are duplicated. The dataset provides a comprehensive perspective, rich for diverse analyses.

<a id='stats-country'></a>

### 2 - <u>EdStatsCountry :</u>

#### 1.1 - Preview of the First and Last Rows of the Dataset
> View of the dataset's start & end, providing a quick data structure glimpse.

In [368]:
print('First 5 rows of the dataset:')
COUNTRY.head()

First 5 rows of the dataset:


Country Code   Short Name   Table Name                     Long Name  \
0          ABW        Aruba        Aruba                         Aruba   
1          AFG  Afghanistan  Afghanistan  Islamic State of Afghanistan   
2          AGO       Angola       Angola   People's Republic of Angola   
3          ALB      Albania      Albania           Republic of Albania   
4          AND      Andorra      Andorra       Principality of Andorra   

  2-alpha code   Currency Unit  \
0           AW   Aruban florin   
1           AF  Afghan afghani   
2           AO  Angolan kwanza   
3           AL    Albanian lek   
4           AD            Euro   

                                       Special Notes  \
0  SNA data for 2000-2011 are updated from offici...   
1  Fiscal year end: March 20; reporting period fo...   
2  April 2013 database update: Based on IMF data,...   
3                                                NaN   
4                                                NaN   

                      Region          Income Group WB-2 code  ...  \
0  Latin America & Caribbean  High income: nonOECD        AW  ...   
1                 South Asia            Low income        AF  ...   
2         Sub-Saharan Africa   Upper middle income        AO  ...   
3      Europe & Central Asia   Upper middle income        AL  ...   
4      Europe & Central Asia  High income: nonOECD        AD  ...   

            IMF data dissemination standard  \
0                                       NaN   
1  General Data Dissemination System (GDDS)   
2  General Data Dissemination System (GDDS)   
3  General Data Dissemination System (GDDS)   
4                                       NaN   

                            Latest population census  \
0                                               2010   
1                                               1979   
2                                               1970   
3                                               2011   
4  2011. Population figures compiled from adminis...   

                             Latest household survey  \
0                                                NaN   
1  Multiple Indicator Cluster Survey (MICS), 2010/11   
2               Malaria Indicator Survey (MIS), 2011   
3       Demographic and Health Survey (DHS), 2008/09   
4                                                NaN   

   Source of most recent Income and expenditure data  \
0                                                NaN   
1            Integrated household survey (IHS), 2008   
2            Integrated household survey (IHS), 2008   
3  Living Standards Measurement Study Survey (LSM...   
4                                                NaN   

  Vital registration complete Latest agricultural census  \
0                         Yes                        NaN   
1                         NaN                    2013/14   
2                         NaN                       2015   
3                         Yes                       2012   
4                         Yes                        NaN   

  Latest industrial data Latest trade data Latest water withdrawal data  \
0                    NaN            2012.0                          NaN   
1                    NaN            2012.0                         2000   
2                    NaN               NaN                         2005   
3                 2010.0            2012.0                         2006   
4                    NaN            2006.0                          NaN   

  Unnamed: 31  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN  

[5 rows x 32 columns]

In [369]:
print('Last 5 rows of the dataset:')
COUNTRY.tail()

Last 5 rows of the dataset:


Country Code    Short Name    Table Name                 Long Name  \
236          XKX        Kosovo        Kosovo        Republic of Kosovo   
237          YEM         Yemen   Yemen, Rep.         Republic of Yemen   
238          ZAF  South Africa  South Africa  Republic of South Africa   
239          ZMB        Zambia        Zambia        Republic of Zambia   
240          ZWE      Zimbabwe      Zimbabwe      Republic of Zimbabwe   

    2-alpha code       Currency Unit  \
236          NaN                Euro   
237           YE         Yemeni rial   
238           ZA  South African rand   
239           ZM  New Zambian kwacha   
240           ZW         U.S. dollar   

                                         Special Notes  \
236  Kosovo became a World Bank member on June 29, ...   
237  Based on official government statistics and In...   
238  Fiscal year end: March 31; reporting period fo...   
239  National accounts data have rebased to reflect...   
240  Fiscal year end: June 30; reporting period for...   

                         Region         Income Group WB-2 code  ...  \
236       Europe & Central Asia  Lower middle income        KV  ...   
237  Middle East & North Africa  Lower middle income        RY  ...   
238          Sub-Saharan Africa  Upper middle income        ZA  ...   
239          Sub-Saharan Africa  Lower middle income        ZM  ...   
240          Sub-Saharan Africa           Low income        ZW  ...   

                IMF data dissemination standard  Latest population census  \
236    General Data Dissemination System (GDDS)                      2011   
237    General Data Dissemination System (GDDS)                      2004   
238  Special Data Dissemination Standard (SDDS)                      2011   
239    General Data Dissemination System (GDDS)                      2010   
240    General Data Dissemination System (GDDS)                      2012   

                               Latest household survey  \
236                                                NaN   
237          Demographic and Health Survey (DHS), 2013   
238  Demographic and Health Survey (DHS), 2003; Wor...   
239          Demographic and Health Survey (DHS), 2013   
240       Demographic and Health Survey (DHS), 2010/11   

    Source of most recent Income and expenditure data  \
236           Integrated household survey (IHS), 2011   
237    Expenditure survey/budget survey (ES/BS), 2005   
238    Expenditure survey/budget survey (ES/BS), 2010   
239           Integrated household survey (IHS), 2010   
240        Integrated household survey (IHS), 2011/12   

    Vital registration complete            Latest agricultural census  \
236                         NaN                                   NaN   
237                         NaN                                   NaN   
238                         NaN                                  2007   
239                         NaN  2010. Population and Housing Census.   
240                         NaN                                   NaN   

    Latest industrial data Latest trade data Latest water withdrawal data  \
236                    NaN               NaN                          NaN   
237                 2006.0            2012.0                         2005   
238                 2010.0            2012.0                         2000   
239                    NaN            2011.0                         2002   
240                    NaN            2012.0                         2002   

    Unnamed: 31  
236         NaN  
237         NaN  
238         NaN  
239         NaN  
240         NaN  

[5 rows x 32 columns]

> <u>Note:</u> The data includes various elements such as dates and region names. Missing values, denoted by 'NaN', are evident. A column named '...' consistently shows repeated and atypical values, all being '...'. Additionally, the 'Unnamed: 31' column appears to be an oversight as it only contains missing values.

#### 1.2 - Columns 

In [370]:
# Display columns names
print(COUNTRY.columns)

Index(['Country Code', 'Short Name', 'Table Name', 'Long Name', '2-alpha code',
       'Currency Unit', 'Special Notes', 'Region', 'Income Group', 'WB-2 code',
       'National accounts base year', 'National accounts reference year',
       'SNA price valuation', 'Lending category', 'Other groups',
       'System of National Accounts', 'Alternative conversion factor',
       'PPP survey year', 'Balance of Payments Manual in use',
       'External debt Reporting status', 'System of trade',
       'Government Accounting concept', 'IMF data dissemination standard',
       'Latest population census', 'Latest household survey',
       'Source of most recent Income and expenditure data',
       'Vital registration complete', 'Latest agricultural census',
       'Latest industrial data', 'Latest trade data',
       'Latest water withdrawal data', 'Unnamed: 31'],
      dtype='object')


> <u>Notes :</u> The data represents metadata pertaining to various countries, encompassing codes, names, economic details, and census dates. The 'Unnamed: 31' column appears redundant and may contain errors or missing values.

#### 1.3 - Missing Values & Percentages
> Overview of missing data in columns, presented as counts & percentages

In [371]:
print("\nNumber of missing values for each column:")
COUNTRY.isnull()


Number of missing values for each column:


Country Code  Short Name  Table Name  Long Name  2-alpha code  \
0           False       False       False      False         False   
1           False       False       False      False         False   
2           False       False       False      False         False   
3           False       False       False      False         False   
4           False       False       False      False         False   
..            ...         ...         ...        ...           ...   
236         False       False       False      False          True   
237         False       False       False      False         False   
238         False       False       False      False         False   
239         False       False       False      False         False   
240         False       False       False      False         False   

     Currency Unit  Special Notes  Region  Income Group  WB-2 code  ...  \
0            False          False   False         False      False  ...   
1            False          False   False         False      False  ...   
2            False          False   False         False      False  ...   
3            False           True   False         False      False  ...   
4            False           True   False         False      False  ...   
..             ...            ...     ...           ...        ...  ...   
236          False          False   False         False      False  ...   
237          False          False   False         False      False  ...   
238          False          False   False         False      False  ...   
239          False          False   False         False      False  ...   
240          False          False   False         False      False  ...   

     IMF data dissemination standard  Latest population census  \
0                               True                     False   
1                              False                     False   
2                              False                     False   
3                              False                     False   
4                               True                     False   
..                               ...                       ...   
236                            False                     False   
237                            False                     False   
238                            False                     False   
239                            False                     False   
240                            False                     False   

     Latest household survey  \
0                       True   
1                      False   
2                      False   
3                      False   
4                       True   
..                       ...   
236                     True   
237                    False   
238                    False   
239                    False   
240                    False   

     Source of most recent Income and expenditure data  \
0                                                 True   
1                                                False   
2                                                False   
3                                                False   
4                                                 True   
..                                                 ...   
236                                              False   
237                                              False   
238                                              False   
239                                              False   
240                                              False   

     Vital registration complete  Latest agricultural census  \
0                          False                        True   
1                           True                       False   
2                           True                       False   
3                          False                       False   
4                          False                        True   
..  

In [372]:
missing_values_table(COUNTRY)

Missing Values  \
Unnamed: 31                                                   241   
National accounts reference year                              209   
Alternative conversion factor                                 194   
Other groups                                                  183   
Latest industrial data                                        134   
Vital registration complete                                   130   
External debt Reporting status                                117   
Latest household survey                                       100   
Latest agricultural census                                     99   
Lending category                                               97   
Special Notes                                                  96   
PPP survey year                                                96   
Source of most recent Income and expenditure data              81   
Government Accounting concept                                  80   
Latest water withdrawal data                                   62   
IMF data dissemination standard                                60   
Balance of Payments Manual in use                              60   
Latest trade data                                              56   
SNA price valuation                                            44   
System of trade                                                41   
National accounts base year                                    36   
Latest population census                                       28   
Income Group                                                   27   
Region                                                         27   
Currency Unit                                                  26   
System of National Accounts                                    26   
2-alpha code                                                    3   
WB-2 code                                                       1   

                                                   Missing Percentages (%)  
Unnamed: 31                                                         100.00  
National accounts reference year                                     86.72  
Alternative conversion factor                                        80.50  
Other groups                                                         75.93  
Latest industrial data                                               55.60  
Vital registration complete                                          53.94  
External debt Reporting status                                       48.55  
Latest household survey                                              41.49  
Latest agricultural census                                           41.08  
Lending category                                                     40.25  
Special Notes                                                        39.83  
PPP survey year                                                      39.83  
Source of most recent Income and expenditure data                    33.61  
Government Accounting concept                                        33.20  
Latest water withdrawal data                                         25.73  
IMF data dissemination standard                                      24.90  
Balance of Payments Manual in use                                    24.90  
Latest trade data                                                    23.24  
SNA price valuation                                                  18.26  
System of trade                                                      17.01  
National accounts base year                                          14.94  
Latest population census                                             11.62  
Income Group                                                         11.20  
Region                                                               11.20  
Currency Unit                                                        10.79  
System of National Accounts                    

In [373]:
# Display missing data > 70%
missing_data_70 = missing_values_table(COUNTRY)[missing_values_table(COUNTRY)['Missing Percentages (%)'] > 70]
print("Display missing data > 70% :")
print(missing_data_70)
print( f"total columns that have missing data > 70% :", missing_data_70.index.size)

Display missing data > 70% :
                                  Missing Values  Missing Percentages (%)
Unnamed: 31                                  241                   100.00
National accounts reference year             209                    86.72
Alternative conversion factor                194                    80.50
Other groups                                 183                    75.93
total columns that have missing data > 70% : 4


> <u>Note :</u> Out of the 32 columns, 4 have over 70% missing data. Notably, the 'Unnamed: 31' column is entirely devoid of data, with a 100% missing rate.

#### 1.4 - Display duplicate rows

In [374]:
duplicated_rows = COUNTRY.duplicated().sum()
print(f"total duplicate rows:  {duplicated_rows}")

total duplicate rows:  0


> <u>Note :</u> We can observe that there are no duplicated rows.

In [375]:
print(f"Number of rows: {COUNTRY.shape[0]}")
print(f"Number of columns {COUNTRY.shape[1]}")
print(f"Number of unique Regions: {COUNTRY['Region'].nunique()}")


Number of rows: 241
Number of columns 32
Number of unique Regions: 7


> <u>Note :</u> The dataset, spanning 241 rows and 32 columns, encompasses 7 unique regions.

#### 1.5 - <u>Notes</u> : 

L'ensemble de données, s'étendant sur 241 lignes et 32 colonnes, contient des métadonnées sur divers pays et 7 régions uniques. Il présente des éléments tels que des dates, des noms de régions et des informations économiques. Plusieurs colonnes ont des valeurs manquantes, notamment la colonne 'Unnamed: 31' qui est entièrement vide. De plus, certaines colonnes affichent des valeurs atypiques.

<a id='data-cleaning'></a>
## Data Cleaning
> Address any issues in the dataset to ensure its quality and reliability. This involves handling missing values, duplicates, and any inconsistencies.

### Remove Unnecessary Columns

In [376]:
# Display missing DATA == 100%
print(missing_values_table(DATA)[missing_values_table(DATA)['Missing Percentages (%)'] == 100])

             Missing Values  Missing Percentages (%)
Unnamed: 69          886930                    100.0


> <u>Remove "Unnamed: 69:"</u> Since you mentioned earlier that this column had 100% missing values, it would make sense to remove it.

In [377]:
# Drop the 'Unnamed: 69' column as it contains 100% missing values

# Calculate the missing values and their percentages for each column
missing_data = missing_values_table(data)

# Identify columns that have 100% missing values
cols_to_drop = missing_data[missing_data['Missing Percentages (%)'] == 100].index.tolist()

# Check if there are any columns to drop
# If you run this code a second time it gives an empty list, this is normal because the column will be deleted here we just have the cad of 'Unnamed: 69'
if cols_to_drop:
    data.drop(cols_to_drop, axis=1, inplace=True)
    plural = 's' if len(cols_to_drop) > 1 else ''
    print(f"{len(cols_to_drop)} column{plural} dropped because it{plural if len(cols_to_drop) == 1 else ' they'} had 100% missing values: {', '.join(cols_to_drop)}")
else:
    print("No columns were dropped as none had 100% missing values.")


1 column dropped because it had 100% missing values: Unnamed: 69


In [378]:
# Display missing data after removing the column 'Unnamed: 69'
print("Display missing data after removing the column")
print(missing_values_table(data))

Display missing data after removing the column
      Missing Values  Missing Percentages (%)
2017          886787                    99.98
2016          870470                    98.14
1973          851385                    95.99
1971          851393                    95.99
1972          851311                    95.98
...              ...                      ...
2011          740918                    83.54
2012          739666                    83.40
2000          710254                    80.08
2005          702822                    79.24
2010          644488                    72.67

[65 rows x 2 columns]


### Duplicated rows
> Remove duplicated rows if any

In [379]:
if duplicated_rows > 0:
    data.drop_duplicates(inplace=True)
    print("Duplicated rows have been removed.")
else:
    print("No duplicated rows found.")

No duplicated rows found.
